### ISSUE TITLE
textlint: Considering RuleError design

### ISSUE THREAD

**Issue by azu** [textlint/textlint#564]
Current `RuleError` design is something wrong.
It cause some error like #244 and confuse rule creator.

```js
report(node, new RuleError(message, paddingObject));
```

We want to get more clean `report` API or `RuleError` API.

## Purposes

- Current computing logic of absolute location + relative location is complex
- `new RuleError(message, { line, column })` is strange logic
	- See next section
- Want to add error `range` for highlight
	- Currently, rule can only report point
	- `fix` has range, but error only report can not use it.
	- ESLint has added `endLine` and `endColumn` for it
	- https://eslint.org/docs/developer-guide/nodejs-api#linterverify

We want to resolve these problems and add `range` feature.

This issue is very related with #244, I think.

## What is RuleError?

`RuleError` receive PaddingObject.
`PaddingObject` has these pattern.

- `{ index }`
- `{ line }`
- `{ line, column }`

`PaddingObject` represent relative value from `node`.

## Basic concept of RuleError

`report(node, ruleError)` accept two value

- `node` is TxtNode
- `ruleError` is instance of `RuleError`

`node` has `loc` and `range` as location variable.
It is **absolute location**.

`ruleError` has `index` or `{ line, column }`.
They is **relative location** from `node`.

## How to calculate absolute position

Intenally, textlint calculate the absolute position of report.

- `text` source
- `node`
- `PaddingObject`

`node`'s `range` and `loc` is absolute position.
`node` + `PaddingObject` is absolute position.

But, can not compose `node`'s `loc` and `{line, column}`.
Because, `{line:1,column:1}` + Padding(`{line:1, column:10}`) may cause break line.
So, textlint need to `text` source to get absolute index from `loc` object.

## What is wrong?

Current `RuleError` design is something wrong.

- Need three separated source (`text`, `node`, `PaddingObject`) to calculate absolute position
	- Actually, `RuleError` receive only `PaddingObject`
	- `new RuleError(message, PaddingObject)`
	- It means that can not calculate absolute position from `ruleError` parmeters.
- `PaddingObject` is `{ line, column }` is inconsist way

Current behavior is following

```js
// when use {line, column}
if (padding.line !== undefined && padding.column !== undefined) {
    if (padding.line > 0) {
        const addedLine = line + padding.line;
        // when report with padding {line, column}, message.column should be 0 + padding.column.
        // In other word, padding line > 0 and message.column start with 0.
        if (padding.column > 0) {
            return {
                line: addedLine,
                column: padding.column // { line: 1, column: 1 } => { line: abs.line + 1, column: 1 } ???
            };
        } else {
            return {
                line: addedLine,
                column
            };
        }
    }
}
```

## Current Use Case

- Receive external {line, column} and pass {line, column}
- https://github.com/textlint-rule/textlint-rule-alex/blob/a93644c5c92b09437a7aa0049b6ffeefd6fe17e4/src/textlint-rule-alex.js#L25-L31
- https://github.com/textlint-rule/textlint-rule-eslint/blob/d5db0e78afc297c74fde736297c7f719871eaac4/src/textlint-rule-eslint.js#L66-L75

```
const reportError = (node, result) => {
    const ruleError = new RuleError(`[${result.ruleId}] ${result.message}`, {
        line: result.line - 1,
        column: result.column - 1
    });
    report(node, ruleError);
};
```

- padding line, padding column -> index
- https://github.com/textlint-ja/textlint-rule-no-doubled-joshi/blob/8fdcf6336539b94ec4fdab0d517b1f7a3220e2dc/src/no-doubled-joshi.js#L133-L150

```js
tokens.reduce((prev, current) => {
    const startPosition = countableTokens.indexOf(prev);
    const otherPosition = countableTokens.indexOf(current);
    // 助詞token同士の距離が設定値以下ならエラーを報告する
    const differenceIndex = otherPosition - startPosition;
    if (differenceIndex <= minInterval) {
        const originalIndex = source.originalIndexFromPosition({
            line: sentence.loc.start.line,
            column: sentence.loc.start.column + (current.word_position - 1)
        });
        // padding positionを計算する
        const padding = {
            index: originalIndex
        };
        report(node, new RuleError(`一文に二回以上利用されている助詞 "${joshiName}" がみつかりました。`, padding));
    }
    return current;
});
```

- padding line, start column
- It may be defectiveness
- https://github.com/textlint-ja/textlint-rule-preset-JTF-style/blob/174187a29631bafcfd7cddadb82e8dbeb43406f8/src/1.1.3.js#L42-L57
- https://github.com/textlint-ja/textlint-rule-preset-JTF-style/blob/174187a29631bafcfd7cddadb82e8dbeb43406f8/src/1.1.5.js#L24-L40

```js
    function reportResult(list, {desumasu, dearu}) {
        list.forEach(({node, matches}) => {
            matches.forEach(match => {
                let message;
                if (desumasu) {
                    message = `本文を常体(である調)に統一して下さい。\n本文の文体は、敬体(ですます調)あるいは常体(である調)のどちらかで統一します。\n"${match.value}"が敬体(ですます調)です。`
                } else if (dearu) {
                    message = `本文を敬体(ですます調)に統一して下さい。\n本文の文体は、敬体(ですます調)あるいは常体(である調)のどちらかで統一します。\n"${match.value}"が常体(である調)です。`
                }
                report(node, new RuleError(message, {
                    line: match.lineNumber - 1,
                    column: match.columnIndex
                }));
            });
        });
    }
```

- padding line, padding column

Realted?: [Maybe Some rule report invalid `line` or `column` · Issue #244 · textlint/textlint](https://github.com/textlint/textlint/issues/244 "Maybe Some rule report invalid `line` or `column` · Issue #244 · textlint/textlint")

(It is bug?)

```js
    function reportResult(list, {desumasu,dearu}) {
        list.forEach(({node, matches}) => {
            matches.forEach(match => {
                let message;
                if (desumasu) {
                    message = `図表のキャプションを敬体(ですます調)に統一して下さい。\n図表のキャプション内で敬体、常体を混在させないことが重要です。\n"${match.value}"が常体(である調)です。`
                } else if (dearu) {
                    message = `図表のキャプションを常体(である調)に統一して下さい。\n図表のキャプション内で敬体、常体を混在させないことが重要です。\n"${match.value}"が敬体(ですます調)です。`
                }
                report(node, new RuleError(message, {
                    line: match.lineNumber - 1,
                    column: match.columnIndex + imagePaddingLet
                }));
            });
        });
    }
```

- self calculate line and column 
- https://github.com/textlint-ja/textlint-rule-max-ten/blob/0eddc7855ab6fcbb7a9b43445b5d5a54a5ed8e76/src/max-ten.js#L29-L35

```js
function addPositions(base, relative) {
    return {
        line: base.line + relative.line - 1, // line 1 + line 1 should be line 1
        column: relative.line == 1 ? base.column + relative.column // when the same line
            : relative.column               // when another line
    };
}
```

## Note 📝

Q. Why does `RuleError` treat relative position?
A. I've written many rule and I get some result.

Almost reporting case was represent `index` + `node.range`.
In other words, Almost rule reporting use `index` from `node`'s start `index`(`range[0]`).

ESLint only use `line` and `column` as absolute position.

- [Node.js API - ESLint - Pluggable JavaScript linter](https://eslint.org/docs/developer-guide/nodejs-api#linterverify)

## Related

- Maybe Some rule report invalid `line` or `column` #244

**Comment by azu** [comment]
I have a idea.

Add `locator` API to `context` object like `fixer` API.

Example.

```js
const reporter = context => {
  const { report, fixer, locator, Syntax, RuleError } = context;
  // locator has many utility 
  // indexOf, locationToIndex, indexToLocation...
  return {
    [Syntax.Str](node){
      report(node, new RuleError("message", {
        location: locator.indexOf("."),
        fix: fixer.insertTextAfter(node, "a")
      };
    }
  };
};
```

**Comment by azu** [comment]
https://github.com/textlint/textlint/releases/tag/v12.2.0
locator is introduced in v12.2.0
